experimenter/experimenter/nimbus-ui/src/components/PageEditBranches/FormBranches/FormFeatureValue/RichFeatureValueEditor.tsx

```diff
@@ -0,0 +1,435 @@
+          severity: "error",
+        },
+      ];
+    }
+
+    // Feature configuration values must be objects.
+    if (rootNode.type !== "Object") {
+      let type: string = rootNode.type;
+      if (rootNode.type === "Literal") {
+        type = typeof rootNode.value;
+      }
+      return [
+        {
+          from: 0,
+          to: view.state.doc.length,
+          message: `Expected a JSON Object, not ${type}`,
+          severity: "error",
+        },
+      ];
+    }
+
+    // We know JSON.parse(text) will succeed because we already parsed text into
+    // an ast above.
+    const validationResult = validate(
+      JSON.parse(text),
+      schema,
+      draft,
+      undefined,
+      false,
+    );
+    if (validationResult.valid) {
+      return [];
+    }
+
+    const diagnostics: Diagnostic[] = [];
+
+    // Keep track of type errors reported by instanceLocation, because a more
+    // specific type error will be followed by less helpful, more generic
+    // errors.
+    const typeErrors = new Set();
+
+    for (const error of validationResult.errors) {
+      let message = error.error;
+
+      if (SCHEMA_RE.exec(message)) {
+        // We will get a more specific error later.
+        continue;
+      }
+
+      const instancePath = error.instanceLocation.split("/").slice(1);
+      let instanceLocation = error.instanceLocation;
+      let nodeKey: keyof FindNodeResult = "value";
+
+      if (error.keyword === "type" && instancePath.length) {
+        typeErrors.add(error.instanceLocation);
+      } else if (error.keyword === ADDITIONAL_PROPERTIES) {
+        const match = ADDITIONAL_PROPERTIES_RE.exec(message);
+        if (match) {
+          const propertyName = match[1];
+
+          // This error gets reported on the object, not the property. Suppress
+          // this error if we've already reported a type error for the property.
+          instanceLocation = `${error.instanceLocation}/${propertyName}`;
+          if (typeErrors.has(instanceLocation)) {
+            continue;
+          }
+
+          typeErrors.add(instanceLocation);
+          instancePath.push(propertyName);
+          message = `Unexpected property "${propertyName}"`;
+          nodeKey = "key";
+        }
+      } else if (
+        IGNORED_ERRORS.includes(message) ||
+        (message === FALSE_BOOLEAN_SCHEMA &&
+          typeErrors.has(error.instanceLocation))
+      ) {
+        continue;
+      }
+
+      const pos = findNode(rootNode, instancePath)[nodeKey]!.loc!;
+      const from = documentPosition(
+        view.state.doc,
+        pos.start.line,
+        pos.start.column - 1,
+      );
+      const to = documentPosition(
+        view.state.doc,
+        pos.end.line,
+        pos.end.column - 1,
+      );
+
+      diagnostics.push({
+        from,
+        to,
+        message,
+        severity: "error",
+      });
+    }
+
+    reportFloatValues(view.state.doc, rootNode, diagnostics);
+
+    return diagnostics;
```

Comment:
I didn't think we'd have to do all this ourselves, I was hoping we could just naively pass the jsonschemas we have straight to this thing and it'd start validating.  But in any case, would it be worth pulling this out of here and into some kind of utility file and testing it separately rather than as part of the component?

Reply:
That is what it does for the most part, with some additional filtering so that we don't get a bunch of repeated error messages.